parl/algorithms/paddle/ddqn.py

```diff
@@ -0,0 +1,78 @@
+#   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import parl
+import paddle
+
+__all__ = ['DDQN']
+
+
+class DDQN(parl.Algorithm):
+    def __init__(self, model, gamma=None, lr=None):
+        """ DQN algorithm
+
+        Args:
+            model (parl.Model): forward neural network representing the Q function.
+            gamma (float): discounted factor for `accumulative` reward computation
+            lr (float): learning rate.
+        """
+        self.model = model
+        self.target_model = copy.deepcopy(model)
+
+        assert isinstance(gamma, float)
+        assert isinstance(lr, float)
+
+        self.gamma = gamma
+        self.lr = lr
+
+        self.mse_loss = paddle.nn.MSELoss(reduction='mean')
+        self.optimizer = paddle.optimizer.Adam(
+            learning_rate=lr, parameters=self.model.parameters())
+
+    def predict(self, obs):
+        """ use self.model (Q function) to predict the action values
+        """
+        return self.model.value(obs)
+
+    def learn(self, obs, action, reward, next_obs, terminal):
+        """ update the Q function (self.model) with DQN algorithm
+        """
+        # Q
+        pred_values = self.model.value(obs)
+        action_dim = pred_values.shape[-1]
+        action = paddle.squeeze(action, axis=-1)
+        action_onehot = paddle.nn.functional.one_hot(
+            action, num_classes=action_dim)
+        pred_value = paddle.multiply(pred_values, action_onehot)
+        pred_value = paddle.sum(pred_value, axis=1, keepdim=True)
+
+        # target Q
+        with paddle.no_grad():
+            greedy_actions = self.model.value(next_obs).argmax(1)
+            g_action_oh = paddle.nn.functional.one_hot(greedy_actions, num_classes=action_dim)
+            max_v = self.target_model.value(next_obs).multiply(g_action_oh).sum(axis=1, keepdim=True)
```

Comment:
Please break down this line of code to make it easier to understand（like line 56~59）